elyra/contents/file_parser.py

```diff
@@ -0,0 +1,279 @@
+    def search_expressions(self) -> Dict[str, List]:
+        # a dictionary of matching regex expressions to use keyed to the json stanza
+        # Parsing for ENVs e.g. search_expressions['NAME_OF_JSON_STANZA'] = ['REGEX_EXP_1', 'REGEX_EXP_2']
+        pass
+    '''
+
+    @abstractmethod
+    def file_as_list_of_code_chunks(self) -> List[str]:
+        # A list of code `chunks` to look through. The size and length of the `chunks` is determined
+        # usually by the language. For now the concrete implementations below is by line-by-line in file
+        pass
+
+    @abstractmethod
+    def get_next_code_chunk(self, filepath):
+        # Implements a generator for lines of code for the specified operation type
+        pass
+
+    @abstractmethod
+    def get_resources(self, filepath):
+        # Returns a model dictionary of all the regex matches for each key in the regex dictionary
+        pass
+
+
+class NotebookFileParser(FileParser):
+
+    _type = "notebook"
+
+    def __init__(self):
+        # super().__init__(operation_filepath)
+        # self.language = None
+        self.parser = None
+
+    def type(self):
+        return self._type
+
+    def file_as_list_of_code_chunks(self, filepath) -> List[str]:
+        code_chunks = []
+        with open(filepath) as f:
+            self.notebook = nbformat.read(f, as_version=4)
+            language = self.notebook['metadata']['kernelspec']['language']
+
+            if language == 'python':
+                self.parser = PythonScriptParser()
+            elif language == 'r':
+                self.parser = RScriptParser()
+
+            for cell in self.notebook.cells:
+                if cell.source and cell.cell_type == "code":
+                    code_chunks.append(cell.source.split('\n'))
+        return [item for subchunk in code_chunks for item in subchunk]
+
+    def get_next_code_chunk(self, filepath):
+        with open(filepath) as f:
+            self.notebook = nbformat.read(f, as_version=4)
+            language = self.notebook['metadata']['kernelspec']['language']
+
+            if language == 'python':
+                self.parser = PythonScriptParser()
+            elif language == 'r':
+                self.parser = RScriptParser()
+
+            for cell in self.notebook.cells:
+                if cell.source and cell.cell_type == "code":
+                    yield cell.source.split('\n')
+
+    def get_resources(self, filepath):
+        model = dict()
+        model["env_list"] = dict()
+        model["inputs"] = dict()
+        model["outputs"] = dict()
+
+        for chunk in self.get_next_code_chunk(filepath):
+            if chunk:
+                # Notebook cells must be further split into individual lines
+                for line in chunk:
+                    matches = self.parser.parse_environment_variables(line)
+                    for key, match in matches:
+                        model[key][match.group(1)] = match.group(2)
+
+        return model
```

Comment:
I believe this method could move into `FileParser` as-is.

I would also make a 'chunk' be a line for now and remove the middle loop.  I.e., we process a "chunk" at a time.

Reply:
I think this makes the most sense implementation-wise. I originally had it like this because we were splitting Notebook cell code on `\n` which got us a list rather than a single string. Do you think I should change `get_next_code_chunk()` for notebooks to return the entire cell as one string (e.g. no `.split('\n')` or do something like the below? 

```
for cell in self.notebook.cells:
    if cell.source and cell.cell_type == "code":
        for line in cell.source.split('\n'):
            yield line
```

Reply:
Sounds good! I'll have Python and R return a list of one item for now and add the type hints